Sources/SwiftGRPCNIO/LengthPrefixedMessageReader.swift

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2019, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import Foundation
+import NIO
+import NIOHTTP1
+
+/// This class reads and decodes length-prefixed gRPC messages.
+///
+/// Messages are expected to be in the following format:
+/// - compression flag: 0/1 as a 1-byte unsigned integer,
+/// - message length: length of the message as a 4-byte unsigned integer,
+/// - message: `message_length` bytes.
+///
+/// - SeeAlso:
+/// [gRPC Protocol](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md)
+internal class LengthPrefixedMessageReader {
+  private var buffer: ByteBuffer!
+  private var state: State = .expectingCompressedFlag
+  private let mode: Mode
+
+  internal init(mode: Mode) {
+    self.mode = mode
+  }
+
+  internal enum Mode { case client, server }
+
+  private enum State {
+    case expectingCompressedFlag
+    case expectingMessageLength
+    case receivedMessageLength(Int)
+    case willBuffer(requiredBytes: Int)
+    case isBuffering(requiredBytes: Int)
+  }
+
+  /// Reads bytes from the given buffer until it is exhausted or a message has been read.
+  ///
+  /// Length prefixed messages may be split across multiple input buffers in any of the
+  /// following places:
+  /// 1. after the compression flag,
+  /// 2. after the message length flag,
+  /// 3. at any point within the message.
+  ///
+  /// - Note:
+  /// This method relies on state; if a message is _not_ returned then the next time this
+  /// method is called it expect to read the bytes which follow the most recently read bytes.
+  /// If a message _is_ returned without exhausting the given buffer then reading a
+  /// different buffer is not an issue.
+  ///
+  /// - Parameter messageBuffer: buffer to read from.
+  /// - Returns: A buffer containing a message if one has been read, or `nil` if not enough
+  ///   bytes have been consumed to return a message.
+  /// - Throws: Throws an error if the compression algorithm is not supported.
+  internal func read(messageBuffer: inout ByteBuffer, compression: CompressionMechanism) throws -> ByteBuffer? {
```

Comment:
The method looks correct to me, but did you consider a slightly different architecture with a "feed" method that "throws as many bytes as possible into the funnel", and the reader calling a delegate method (e.g. `messageAvailable`) whenever it finishes reading a full message? That could help avoid having to call `read` in a loop. Might be slightly less efficient than the current approach but easier to reason about, but not sure — haven't thought that through.

Optional (I'm not sure myself which approach is better); let me know what you think.

Reply:
I didn't; this just grew naturally out of the code for the server. The read in a loop does feel quite strange, but pushing the loop into another method within the reader, e.g. `consume(messageBuffer:compression:) -> [ByteBuffer]`  would clear things up at the call site.

Reply:
So I guess we keep this as is?

Reply:
I added the `consume(messageBuffer:compression) -> [ByteBuffer]` API to improve readability.

I suppose the main difference between the two approaches is that feeding in all bytes would require you to copy all the bytes into an internal buffer every time. With the current approach we only do that if a message is not contained within one `ByteBuffer`. The only trade off is that the "feed" approach would handle the message length being split across `ByteBuffer`s.

I'm good with either and happy to change it. What do you think?

Reply:
Why would we need to copy all the bytes into an internal buffer every time? My understanding would be that we'd try to go over whatever we get passed in, and whenever we encounter a full message in the input, call the "processOutput" "delegate" method right there. So we'd only need to store whatever "half message" is left in the input buffer once all "full messages" have been processed.

However, I think the current approach is fine, so this is not worth changing.

Reply:
Sorry, I just realised you were suggesting two changes:

> ... a "feed" method that "throws as many bytes as possible into the funnel"

I understood this to be "consume all of the bytes we can, even if we can't produce a message", which we would have to write into a buffer. As opposed to "read as many _usable_ bytes as we can" which we're doing now (i.e we can still have bytes leftover if the message length is split across buffers).

> the reader calling a delegate method (e.g. `messageAvailable`) whenever it finishes reading a full message

This could go with either approach. I think the recent change returning `[ByteBuffer]` is fine here.